### PR TITLE
Fixed `FiniteSet.evalf(...)` with a subs argument

### LINKED ISSUE #20291: [bug] subs option to evalf isn't working for sets
Evalf does not seem to work on solveset. 

Code to replicate:
```
from sympy.stats import Uniform, cdf
from sympy import Symbol, Eq, solveset, solve

x = Symbol("x")
a = Symbol("a")
b = Symbol("b")

D = Uniform("d", a, b)
F_x = cdf(D)(x)

u = Symbol("u")
expr = solveset(Eq(F_x, u), x)
expr.evalf(subs={u:0.6, a: 0, b: 2})
>> returns {a - u*(a - b)}
```

If instead I use solve, it works:

```
expr = solve(Eq(F_x, u), x)[0]
expr.evalf(subs={u:0.6, a: 0, b: 2})
>> returns 1.20000000000000
```

### PR DESCRIPTION
**Pull request by Maelstrom6**
#### References to other Issues or PRs

Fixes #20291 

#### Brief description of what is fixed or changed

`FiniteSet(a).evalf(subs={a: 1})` would return `FiniteSet(a)` instead of `FiniteSet(1)`. This PR fixes this and similar issues.

#### Release Notes

* sets
    * `FiniteSet.evalf` with a subs argument now does the substitution. Previously, the argument was ignored.

### CODE REVIEW

File: sympy/core/evalf.py [hunk 1]
```diff
@@ -1473,7 +1473,11 @@ def evalf(self, n=15, subs=None, maxn=100, chop=False, strict=False, quad=None,
             result = evalf(self, prec + 4, options)
         except NotImplementedError:
             # Fall back to the ordinary evalf
-            v = self._eval_evalf(prec)
+            from sympy.core.basic import Basic
+            if isinstance(self, Basic) and subs is not None:  # issue 20291
```
Comment (asmeurer) [inline sympy/core/evalf.py:1477]: What is the purpose of this change? Won't `self` always be `Basic`?

**Comment by asmeurer** [comment]
Changes like this should have release notes entries.

File: sympy/core/evalf.py (hunk 1 above)
Comment (Maelstrom6) [reply sympy/core/evalf.py:1477]: `EvalfMixin` is not an instance of `Basic`. I think all its children are instances of `Basic` but I am not 100% sure. That's why I added it in. Should I remove the `Basic` check?

**Comment by oscarbenjamin** [comment]
I notice that other set classes don't have .evalf or it doesn't work properly e.g.:
```
In [116]: x, a, b, c = symbols('x, a, b, c')                                                                                                                  

In [117]: S = solveset(1 + b/x + pi/x**2, x, Reals)                                                                                                           

In [118]: S.evalf(subs={b:1})                                                                                                                                 
...
AttributeError: 'EmptySet' object has no attribute '_eval_evalf'
```
It looks like EmptySet, ConditionSet and Intersection don't work properly with .evalf.

**Comment by Maelstrom6** [comment]
I'll fix `EmptySet` and `Integers` in this PR. I could not replicate the problem with `Complement` and `Intersection`. I'll try `ConditionSet` in a few hours.

**Comment by oscarbenjamin** [comment]
Take a look at #20356

**Comment by Maelstrom6** [comment]
Looks like `Interval`, `Union`, `Complement`, `FiniteSet` all inherit from `EvalfMixin`. But `Set` does not do so. This is one of the reasons why `evalf` does not work so well on anything relating to sets. Should I move the inheritance up to `Set`? 

I also feel that `Singleton` should also have something like 
```python
def _eval_evalf(self, prec):
    return self
```
This way we don't have to rewrite the same thing for `EmptySet`, `Reals`, `Complexes`, etc.

**Comment by oscarbenjamin** [comment]
Moving `EvalfMixin` up to `Set` makes sense. The mixin defines `_eval_evalf`.

**Comment by Maelstrom6** [comment]
I think this adds more functionality that wasn't originally intended when sets were first designed. There are very few tests on these kinds of things and I'm sure other classes in sets.py need to override `eval_evalf`. I don't have much time to think of good tests, but I will add `eval_evalf` to the other classes in a few days.

**Comment by Maelstrom6** [comment]
Note that I took `_eval_evalf` from `FiniteSet` and from `Union` and I took the best parts of both. So we rather raise an error. And we rather use `_evalf` rather than `_eval_evalf`, since the former returns `self` upon failure while the latter returns `None`.

I'm still not sure on what tests to add. The only types of tests I can think of are of the form
```python
# example for SymmetricDifference test
from sympy import *
x = symbols('x')
expr = SymmetricDifference(FiniteSet(x), FiniteSet(0))
assert expr.evalf(subs={x: 0}) == EmptySet
```

Any tests test advice much appreciated.

**Comment by Maelstrom6** [comment]
The last 2 pushes have both failed but neither were through a fault of my own.

Any code review or testing advice is much appreciated.

File: sympy/sets/sets.py [hunk 2]
```diff
@@ -794,6 +794,12 @@ def as_relational(self, *symbols):
                 'number of symbols must match the number of sets')
         return And(*[s.as_relational(i) for s, i in zip(self.sets, symbols)])
 
+    def _eval_evalf(self, prec):
+        try:
+            return self.func(*[arg._evalf(prec) for arg in self.args])
+        except (TypeError, ValueError, NotImplementedError):
+            raise TypeError("Not all sets are evalf-able")
```
Comment (oscarbenjamin) [inline sympy/sets/sets.py:801]: I would like to limit exception catching as much as possible because it often ends up masking bugs. Is there a need to catch all of these?

File: sympy/sets/sets.py [hunk 3]
```diff
@@ -1577,6 +1583,12 @@ def as_relational(self, symbol):
 
         return And(A_rel, B_rel)
 
+    def _eval_evalf(self, prec):
+        try:
+            return self.func(*[arg._evalf(prec) for arg in self.args])
+        except (TypeError, ValueError, NotImplementedError):
+            raise TypeError("Not all sets are evalf-able")
```
Comment (oscarbenjamin) [inline sympy/sets/sets.py:1590]: Is this not the same as the method inherited from the superclass?

Comment (oscarbenjamin) [reply sympy/sets/sets.py:1590]: Oh I see the method isn't defined on `Set`.

File: sympy/sets/sets.py [hunk 4]
```diff
@@ -1645,6 +1657,9 @@ def _contains(self, other):
     def as_relational(self, symbol):
         return false
 
+    def _eval_evalf(self, prec):
+        return self
```
Comment (oscarbenjamin) [inline sympy/sets/sets.py:1661]: Is it necessary to provide this implementation if the superclass has not changed this method from the one inherited from `EvalfMixin`?

File: sympy/sets/sets.py (hunk 2 above)
Comment (Maelstrom6) [reply sympy/sets/sets.py:801]: I don't think it is nessesary. I'll remove them.

File: sympy/sets/sets.py (hunk 3 above)
Comment (Maelstrom6) [reply sympy/sets/sets.py:1590]: Maybe we should define it on `Set` and then only change it for things like `EmptySet` and `UniversalSet`.

File: sympy/sets/sets.py (hunk 4 above)
Comment (Maelstrom6) [reply sympy/sets/sets.py:1661]: I think it is nessesary because the default `_eval_evalf` returns `None`. Some classes (like `Union`) incorrectly use `_eval_evalf` on its arguments when it should use `_evalf`. The former returns `None` by default while the latter returns `self`.

Maybe I should have changed `Union` instead of `EmptySet`.

File: sympy/sets/sets.py (hunk 3 above)
Comment (oscarbenjamin) [reply sympy/sets/sets.py:1590]: Does the same implementation not work for `EmptySet` which has empty args?

File: sympy/core/evalf.py (hunk 1 above)
Comment (sidhu1012) [inline sympy/core/evalf.py:1477]: I think it should not be isinstance(..., Basic), because Classes in Physics module are not Basic and .evalf() is being added to them.

It should rather be 

if subs is not None:
    try:
        v = self.subs(subs)._eval_evalf(prec)
    except Error:
        .......

Comment (Maelstrom6) [reply sympy/core/evalf.py:1477]: Will do. I know @oscarbenjamin does not really like try catching so I think I'll do `hasattr('_evalf')` instead.

Comment (sidhu1012) [reply sympy/core/evalf.py:1477]: Okay!

File: sympy/sets/sets.py (hunk 3 above)
Comment (Maelstrom6) [reply sympy/sets/sets.py]: It does. Thanks. I had no clue that the same implementation worked.

**Comment by oscarbenjamin** [comment]
I think this looks good and we could use it.

The change in the evalf is really a workaround for the fact that we can't pass the full set of evalf options to `_eval_evalf` though. The bigger problem is that the design of `_eval_evalf` seems to be broken: https://github.com/sympy/sympy/issues/20379#issuecomment-730563821

File: sympy/sets/sets.py [hunk 5]
```diff
@@ -635,6 +635,9 @@ def _boundary(self):
     def _measure(self):
         raise NotImplementedError("(%s)._measure" % self)
 
+    def _eval_evalf(self, prec):
+        return self.func(*[arg._evalf(prec) for arg in self.args])
```
Comment (sidhu1012) [inline sympy/sets/sets.py:639]: I would suggest this as it would keep #20379 fixed.

return self.func(*[arg.evalf(n=prec_to_dps(prec)) for arg in self.args])

**Comment by oscarbenjamin** [comment]
This has conflicts now

**Comment by Maelstrom6** [comment]
I'll fix that and add `.evalf` around the beginning of December. I'm a bit short for time at the moment.

**Comment by Maelstrom6** [comment]
Any further suggestions?

File: sympy/core/tests/test_evalf.py [hunk 6]
```diff
@@ -579,3 +579,16 @@ def test_issue_13425():
...
+
+
+def test_issue_20291():
+    from sympy import FiniteSet, Complement, Intersection, Reals, EmptySet
+    a = Symbol('a')
+    b = Symbol('b')
+    A = FiniteSet(a, b)
+    assert A.evalf(subs={a: 1, b: 2}) == FiniteSet(1, 2)
+    B = FiniteSet(a-b, 1)
+    assert B.evalf(subs={a: 1, b: 2}) == FiniteSet(-1, 1)
```
Comment (oscarbenjamin) [inline sympy/core/tests/test_evalf.py:591]: I think that the subs arguments should be evalfed to give Float e.g.:
```
In [9]: (x**y).evalf(subs={y:1, x:1})
Out[9]: 1.00000000000000
```
Is that not what happens with other evalf substitutions?

Comment (Maelstrom6) [reply sympy/core/tests/test_evalf.py:591]: It does return a float when run but it's not a problem. I was following the design of this:
https://github.com/sympy/sympy/blob/e036b1cfa1157b796b19ed0a1a0af97d40269366/sympy/core/tests/test_evalf.py#L505-L510

Comment (oscarbenjamin) [reply sympy/core/tests/test_evalf.py:591]: Oh, okay. Well I think if the output is a Float then the test should show a Float. Ideally Float and Integer wouldn't compare equal like this (see #20033 where it looks like I had to change this test).

**Comment by Maelstrom6** [comment]
It's been a while, but are there any further suggestions?

**Comment by oscarbenjamin** [comment]
I think this looks good. Thanks!
